src/main/java/com/puppycrawl/tools/checkstyle/filters/SuppressWithNearbyTextFilter.java

```diff
@@ -0,0 +1,623 @@
+        suppressions.clear();
+
+        for (int lineNo = 0; lineNo < fileText.size(); lineNo++) {
+            final Optional<Suppression> suppression = getSuppression(fileText, lineNo);
+            suppression.ifPresent(suppressions::add);
+        }
+    }
+
+    /**
+     * Tries to extract the suppression from the given line.
+     *
+     * @param fileText {@link FileText} instance.
+     * @param lineNo line number.
+     * @return {@link Optional} of {@link Suppression}.
+     */
+    private Optional<Suppression> getSuppression(FileText fileText, int lineNo) {
+        final String line = fileText.get(lineNo);
+        final Matcher nearbyTextMatcher = nearbyTextPattern.matcher(line);
+
+        Suppression suppression = null;
+        if (nearbyTextMatcher.find()) {
+            final String text = nearbyTextMatcher.group(0);
+            suppression = new Suppression(text, lineNo + 1, this);
+        }
+
+        return Optional.ofNullable(suppression);
+    }
+
+    /**
+     * Finds the nearest {@link Suppression} instance which can suppress
+     * the given {@link AuditEvent}. The nearest suppression is the suppression which scope
+     * is before the line and column of the event.
+     *
+     * @param suppressions collection of {@link Suppression} instances.
+     * @param event {@link AuditEvent} instance.
+     * @return {@link Suppression} instance.
+     */
+    private static Optional<Suppression> getNearestSuppression(Collection<Suppression> suppressions,
+                                                               AuditEvent event) {
+        return suppressions
+                .stream()
+                .filter(suppression -> suppression.isMatch(event))
+                .findFirst();
+    }
+
+    /** The class which represents the suppression. */
+    private static class Suppression {
+
+        /** The first line where warnings may be suppressed. */
+        private final int firstLine;
+
+        /** The last line where warnings may be suppressed. */
+        private final int lastLine;
+
+        /** The regexp which is used to match the event source.*/
+        @XdocsPropertyType(PropertyType.PATTERN)
+        private final Pattern eventSourceRegexp;
+
+        /** The regexp which is used to match the event message.*/
+        @XdocsPropertyType(PropertyType.PATTERN)
+        private Pattern eventMessageRegexp;
+
+        /** The regexp which is used to match the event ID.*/
+        @XdocsPropertyType(PropertyType.PATTERN)
+        private Pattern eventIdRegexp;
+
+        /**
+         * Constructs new {@code Suppression} instance.
+         *
+         * @param text suppression text.
+         * @param lineNo suppression line number.
+         * @param filter the {@code SuppressWithNearbyTextFilter} with the context.
+         * @throws IllegalArgumentException if there is an error in the filter regex syntax.
+         */
+        /* package */ Suppression(
```

Comment:
I believe these package comments aren't needed anymore since we are Java 11.

Reply:
Getting this without it:
```
[INFO] PMD Failure: com.puppycrawl.tools.checkstyle.filters.SuppressWithNearbyTextFilter$Suppression:495 Rule:CommentDefaultAccessModifier Priority:3 To avoid mistakes add a comment at the beginning of the Suppression constructor if you want a default access modifier.

```

Reply:
Yes, you should add `private` to the constructor since no one uses it outside the class.

Reply:
Done.